src/TableOperations.jl

```diff
@@ -410,4 +410,42 @@ Base.eltype(x::RowPartitions) = Vector{MaterializedRow}
     return resize!(v, i), y === nothing ? nothing : (y[2],)
 end
 
+struct NarrowTypes{T} <: Tables.AbstractColumns
+    x::T
+    schema::Tables.Schema
+end
+
+schema(nt::NarrowTypes) = getfield(nt, :schema)
+
+narrowarray(x) = mapreduce(typeof, Base.promote_typejoin, x)
+narrowarray(::Type{T}, x::AbstractArray{T}) where {T} = x
+narrowarray(::Type{T}, x::AbstractArray{S}) where {T, S} = Vector{T}(x)
+
+"""
+    Tables.narrowtypes(source) => Tables.NarrowTypes
+    source |> Tables.narrowtypes() => Tables.NarrowTypes
+
+Take a Tables.jl-compatible source, with potentially "wide" column types, and re-infer the schema by
+promoting the types of each actual value for each column. Useful, for example, when a columnar table source
+has a column type of `Any`, and a more concrete type is desired. Uses `Base.promote_typejoin` internally
+to do actual type promotion.
```

Comment:
Note that `promote_typejoin` is just `typejoin` with special handling for `Missing` and `Nothing`, so speaking about promotion is a bit misleading. Maybe using `promote` be default would be more useful? For example if you have a mix of `Int` and `Float64`, `promote_typejoin` will give `Real`, which isn't ideal.